File: [1]Tutorial_for_Python_and_R/4_polynomial_linear_regression.py
# -*- coding: utf-8 -*-

# Polynomial Regression

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

import os

# Set the directory
currentPath = os.getcwd()
print(currentPath)
os.chdir('C:\\Users\\KYY\\Desktop\\ML_Ato_Z\\[0]Data')


# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
# make X -> matrix, y -> vector
X = dataset.iloc[:, 1:2].values   # don't need 'Position' column  # vector -> matrix
y = dataset.iloc[:, -1].values


# No - split data (-> very small data )

# No - feature scaling
#    the library we'll use do feature scaling


# Fitting Linear Regression to the dataset
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X, y)

# Fitting Polynomial Regression to the dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4)   # transform X metrix feature to polynomial
X_poly = poly_reg.fit_transform(X)   # matrix X -> create columns of polynomial
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, y)


# Visualizing the Linear Regression results
plt.scatter(X, y, color = 'red')
plt.plot(X, lin_reg.predict(X), color = 'blue')
plt.title('Linear Regression')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

# Visualizing the Polynomial Regression results
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape(len(X_grid), 1)

plt.scatter(X, y, color = 'red')
plt.plot(X_grid, lin_reg_2.predict(poly_reg.fit_transform(X_grid)), color = 'blue')
plt.title('Polynomial Regression')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()


# Predicting a new result with Linear Regression
lin_reg.predict(6.5)   # salary or level 6.5

# Predicting a new result with polynomial Regression
lin_reg_2.predict(poly_reg.fit_transform(6.5))  # salary or level 6.5

File: tutorial_for_Python_and_R/4_polynomial_linear_regression.R
# Polynomial Regression

# Set the directory
setwd('C:/Users/KYY/Desktop/ML_Ato_Z/[0]Data')

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Feature Scaling - included in Linear regression package


# Fitting Linear Regression to the dataset
?lm
lin_reg = lm(formula = Salary ~ ., 
             data = dataset)
summary(lin_reg)


# Fitting Polynomial Regression to the dataset
dataset$Level2 = dataset$Level^2 # create new column
dataset$Level3 = dataset$Level^3 # create new column
dataset$Level4 = dataset$Level^4 # create new column
poly_reg = lm(formula = Salary ~ ., 
              data = dataset)
summary(poly_reg)



# Visualizing the Linear Regression results
install.packages('ggplot2')
library(ggplot2)
ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') + 
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
            colour = 'blue') + 
  ggtitle('Truth or Bluff (Linear Regression)') + 
  xlab('Level') + 
  ylab('Salary')


# Visualizing the Polynomial Regression results
ggplot() + 
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') + 
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = dataset)),
            colour = 'blue') + 
  ggtitle('Truth or Bluff (Polynomial Regression)') + 
  xlab('Level') + 
  ylab('Salary')


# Predicting a new result with Linear Regression
data.frame(Level = 6.5)
y_pred = predict(lin_reg, newdata = data.frame(Level = 6.5))
y_pred
?predict

# Predicting a new result with Polynomial Regression
y_pred_poly = predict(poly_reg, newdata = data.frame(Level = 6.5, Level2 = 6.5^2, Level3 = 6.5^3, Level4 = 6.5^4))
y_pred_poly
